Add GraniteMoeHybrid support for 4.0

Ssukriti opened the pull request:
# What does this PR do?

The PR adds support for upcoming Granite4.0 models. It terms of model architecture, it is a hybrid class with shared MLP layer and Bamba layers.


Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

Ssukriti committed:
initial config and MLA layer

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
first pass at decoder

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
completion of layers

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
modeling class

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
adding hybrid class to imports

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix imports granitemoehybrid

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix granitehybrid imports

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix granitehybrid import

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix generated modeling file

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
add some comments

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
minor fixes in layers

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
add sharedMLP layer

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
correct layer names

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fixes in mamba config

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix mamba config

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
change name of MLP layer

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix seq mizer layers

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
correct mamba config

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fixes in param names

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
enable hybrid model

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
update config

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix config granite hybrid

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix attention layer

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
cleanup to re-use mamba code

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
keep layer types

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
attention bias cleanup

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
update mamba layer name

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
first pass at tests

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
first pass at tests

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
use granite attention

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
fix: self attn weights

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
pass at making pos_emb optional

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
initialize self_attn only as needed

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Ssukriti committed:
overwrite forward to create HybridMambaCache

Signed-off-by: Sukriti-Sharma4 <sukriti.sharma4@ibm.com>

Rocketknight1 commented:
cc @arthurzucker for text models!

vasqu commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,505 @@
+class GraniteMoeHybridSdpaAttention(GraniteMoeSharedSdpaAttention):
+    pass
+
+GRANITEMOEHYBRID_ATTENTION_CLASSES = {
```
Just as a heads up, I think it would be nice to follow using the new attention interface (see #35235 for the original PR). Llama can also provide a good first pointer for this, e.g. https://github.com/huggingface/transformers/blob/de182ba2690fe6c3466f6463c7f4b3a61694b885/src/transformers/models/llama/modeling_llama.py#L216

(Except I'm missing that this is a more special kind of attention here :D )

alex-jw-brooks replied:
Thanks for the heads up @vasqu! We are still cleaning up this branch a bit, will take a look at this once the tests are in a better state 🙂

alex-jw-brooks replied:
Thanks for the pointer @vasqu! Refactored this PR to the new attention interface 😄

alex-jw-brooks committed:
Log invalid layer types

alex-jw-brooks committed:
Add attention outputs test

alex-jw-brooks committed:
Only emit attentions/logits if not None

alex-jw-brooks committed:
Fix config test hidden size divisibility

alex-jw-brooks committed:
mark granitmoehybrid as stateful

alex-jw-brooks committed:
Initialize mamba convolutional layers

alex-jw-brooks committed:
Formatting fixes

alex-jw-brooks committed:
config docstring, removed some unused attrs

alex-jw-brooks committed:
Fix missing arg in models test

alex-jw-brooks committed:
Fix create and check decoder model test

alex-jw-brooks committed:
support logits to keep in granitemoe

alex-jw-brooks committed:
regen to pass logits_to_keep

alex-jw-brooks committed:
Allow None or rope

alex-jw-brooks committed:
Fix gradient checkpointing

alex-jw-brooks committed:
Add granitemoehybrid as special cache for generate check

alex-jw-brooks committed:
Remove unused MLA refs

alex-jw-brooks committed:
Fix mamba layer mask

alex-jw-brooks committed:
Remove logits to keep from config

alex-jw-brooks committed:
Minor docstring nits

alex-jw-brooks committed:
Update licenses

alex-jw-brooks committed:
Enable cache by default

alex-jw-brooks committed:
map layer types to layer block type

alex-jw-brooks committed:
First pass at granite moe hybrid docs

alex-jw-brooks committed:
Ignore granite moe hybrid in valid checkpoint check

alex-jw-brooks committed:
Align attention interfaces

alex-jw-brooks committed:
regenerate modular granitemoeshared attention interface

alex-jw-brooks committed:
Align granite moe hybrid attn interface

alex-jw-brooks committed:
run formatting

alex-jw-brooks committed:
Handle mamba initialization

ArthurZucker commented on src/transformers/models/granitemoe/modeling_granitemoe.py:
nice updates

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,528 @@
+                position_embeddings=position_embeddings,
+                **kwargs,
+            )
+        else:
+            raise ValueError(f"Expected layer type in ['attention', 'mamba'], got {self.layer_type}")
```
config check should be sufficient!

alex-jw-brooks replied:
Good point! Moved it 😄

ArthurZucker replied:
still todo 😉

alex-jw-brooks replied:
Whoops, sorry! Fixed

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
Very nice! Thanks for highlighting the key differences!

ArthurZucker reviewed (commented):
Its already in a great state! Thanks a lot both 🤗

ArthurZucker commented:
Feel free to ping me again!

alex-jw-brooks committed:
avoid conditional attr defs

alex-jw-brooks committed:
Move hybrid layer validation to config

alex-jw-brooks committed:
Add placeholder integration tests

alex-jw-brooks committed:
Docs nits / Update model names

alex-jw-brooks commented:
Thanks @ArthurZucker! It's ready for another look when you get the chance!

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,526 @@
+
+        hidden_states = self.input_layernorm(hidden_states)
+        self_attn_weights = None
+        if self.layer_type == "mamba":
```
I am thinking let's remove the check on type, rely rather on the check of self.self_attn is not None?

alex-jw-brooks replied:
I agree, I also didn't like `self.mamba` being conditionally undefined. Updated this to define both in `__init__` and just check do mama if `self.mamba is not None`  and attention otherwise 🙂

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,526 @@
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        moe_hidden_states, router_logits = self.block_sparse_moe(hidden_states)
+
+        if self.shared_mlp is None:
```
I don't know if you answered or not, is there two different checkpoint being released, one with / one without this?

alex-jw-brooks replied:
The models that are about to come out do use it! I think there are likely experiments ongoing without it, but am not sure about concrete plans for when they'll be released since I'm not the one training the models 🙂

ArthurZucker replied:
In that case lets remove what's uncertain! 🤗

alex-jw-brooks replied:
Sounds good! Removed the case with 0 experts, I'll open a follow-up PR if it ends up being used in a model to be released 😄

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,526 @@
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
+                    decoder_layer.__call__,
+                    hidden_states,
+                    layer_mask,
+                    past_key_values,
+                    output_attentions,
+                    use_cache,
+                    cache_position,
+                    output_router_logits,
+                    position_embeddings,
+                )
+            else:
+                layer_outputs = decoder_layer(
+                    hidden_states,
+                    attention_mask=layer_mask,
+                    past_key_value=past_key_values,
+                    output_attentions=output_attentions,
+                    use_cache=use_cache,
+                    cache_position=cache_position,
+                    output_router_logits=output_router_logits,
+                    position_embeddings=position_embeddings,
+                )
```
let's use the new GradientCHeckpointingLayer wdyt?

alex-jw-brooks replied:
Definitely, that is a lot cleaner! I updated the models in the chain for modular to all use the gradient checkpointing layer (GraniteMoe/GraniteMoeShared/GraniteMoeHybrid)

ArthurZucker commented on src/transformers/models/granitemoehybrid/modular_granitemoehybrid.py:
```diff
@@ -0,0 +1,526 @@
+
+        next_cache = next_decoder_cache if use_cache else None
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, next_cache, all_hidden_states, all_self_attns] if v is not None)
```
we have a `@can_return_tuple` for the forward
```suggestion
```

alex-jw-brooks replied:
Nice! Added

ArthurZucker commented on tests/models/granitemoehybrid/test_modeling_granitemoehybrid.py:
```diff
@@ -0,0 +1,528 @@
+    )
+
+
+class GraniteMoeHybridModelTester:
```
can we try to inherit tests from closes model so mambda in the same fashion as here https://github.com/huggingface/transformers/blob/86d38f163fb453bf8e0e878316c7217143a6d1e2/tests/models/gemma2/test_modeling_gemma2.py#L51

alex-jw-brooks replied:
Good idea! The closest models are for the tests are `bamba`. Consolidated a bit to use Bamba tests, should be way easier to look at now 🤞

ArthurZucker replied:
perfect!

ArthurZucker reviewed (approved):
Very nice use of modular thanks a lot! 🤗

alex-jw-brooks committed:
Clean up forward conditions

alex-jw-brooks committed:
Use gradient checkpointing layer

alex-jw-brooks committed:
Remove some copied bamba tests + inherit

align test init

delete more tests

Use common layer init with bamba tests

finish test consolidation

berserkr reviewed (commented):
https://github.com/huggingface/transformers/blob/8274d2c5e5ff4d1cebb9f60c72e3b0d417726b76/src/transformers/models/granitemoehybrid/modeling_granitemoehybrid.py#L1199 - `std` initialized twice - std = self.config.initializer_range

alex-jw-brooks committed:
avoid redundant intermediate std var

alex-jw-brooks committed:
use @can_return_tuple

alex-jw-brooks committed:
Remove unused moe state

alex-jw-brooks commented:
Thanks @berserkr! There were two because of `modular` expanding the superclass implementation that also set it. Updated to just pass the config value directly so it's less weird looking 🙂

alex-jw-brooks committed:
make skipped test names consistent

alex-jw-brooks commented:
Thank you very much for the fast review @ArthurZucker! I've made all the changes 🙂

alex-jw-brooks committed:
Fix docstring order

ArthurZucker reviewed (approved):
Marvelous ! Merging once the build PR passes! (should be easy to fix!)

alex-jw-brooks committed:
Add missing toc

alex-jw-brooks committed:
Always create the shared mlp

alex-jw-brooks committed:
Fix name in docstring

alex-jw-brooks commented:
Thanks @ArthurZucker! Added the missing TOC entry and removed the currently unused shared condition for the MLP, should pass now! 🤞

alex-jw-brooks committed:
link preview model in docs
